src/layout/msWorkspace/msWindow.ts

```diff
@@ -189,6 +199,25 @@ export class MsWindow extends Clutter.Actor {
 
         this.lifecycleState = lifecycleState;
         this.app = app;
+        this.createdAt = createdAt;
+        this.trackAppChanges();
+        this.connect('app-changed', () => {
+            this.trackAppChanges();
+            if (
+                this.lifecycleState.type === 'window' ||
+                this.lifecycleState.type === 'app-placeholder'
+            ) {
+                this.lifecycleState.matchingInfo.appId = this.app.id;
+            }
+            if (this.msContent.contains(this.placeholder)) {
+                this.msContent.remove_child(this.placeholder);
+            }
+            this.placeholder = this.buildPlaceHolder();
+            this.msContent.placeholder = this.placeholder;
+            if (this.lifecycleState.type === 'app-placeholder') {
```

Comment:
Can this ever happen? The app wouldn't ever change if it is in the placeholder state, would it?
If you agree, maybe we should assert that at the top of the callback?